presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java

```diff
@@ -0,0 +1,210 @@
+                                        TEST_TABLE_HANDLE,
+                                        ImmutableList.of(inputSymbol),
+                                        ImmutableMap.of(inputSymbol, inputColumnHandle)));
+                    })
+                    .withSession(MOCK_SESSION)
+                    .matches(project(
+                            ImmutableMap.of("expr_dereference", expression(new SymbolReference(projectedColumnName))),
+                            tableScanWithProvidedHandles(expectedTableHandle, TupleDomain.all(), ImmutableMap.of(projectedColumnName, projectedColumnHandle))));
+        }
+        finally {
+            ruleTester.close();
+        }
+    }
+
+    private MockConnectorFactory createMockFactoryWithoutApplyProjection(Map<String, ColumnHandle> assignments)
+    {
+        List<ColumnMetadata> metadata = assignments.entrySet().stream()
+                .map(entry -> new ColumnMetadata(entry.getKey(), ((TpchColumnHandle) entry.getValue()).getType()))
+                .collect(Collectors.toList());
+
+        return MockConnectorFactory.builder()
+                .withListSchemaNames(connectorSession -> ImmutableList.of(TEST_SCHEMA))
+                .withListTables((connectorSession, schema) -> TEST_SCHEMA.equals(schema) ? ImmutableList.of(TEST_SCHEMA_TABLE) : ImmutableList.of())
+                .withGetColumns(schemaTableName -> schemaTableName.equals(TEST_SCHEMA_TABLE) ? metadata : ImmutableList.of())
+                .build();
+    }
+
+    private MockConnectorFactory createMockFactoryWithApplyProjection(
+            List<ConnectorExpression> inputProjections,
+            Map<String, ColumnHandle> inputAssignments,
+            ConnectorTableHandle applyProjectionTableHandle,
+            List<ConnectorExpression> applyProjectionProjections,
+            List<Assignment> applyProjectionAssignments)
+    {
+        List<ColumnMetadata> metadata = inputAssignments.entrySet().stream()
+                .map(entry -> new ColumnMetadata(entry.getKey(), ((TpchColumnHandle) entry.getValue()).getType()))
+                .collect(Collectors.toList());
+
+        return MockConnectorFactory.builder()
+                .withListSchemaNames(connectorSession -> ImmutableList.of(TEST_SCHEMA))
+                .withListTables((connectorSession, schema) -> TEST_SCHEMA.equals(schema) ? ImmutableList.of(TEST_SCHEMA_TABLE) : ImmutableList.of())
+                .withGetColumns(schemaTableName -> schemaTableName.equals(TEST_SCHEMA_TABLE) ? metadata : ImmutableList.of())
+                .withApplyProjection((session, handle, projections, assignments) -> {
+                    // Assert that the input is expected
+                    assertEquals(projections, inputProjections);
+                    assertEquals(assignments, inputAssignments);
+
+                    return Optional.of(new ProjectionApplicationResult<>(applyProjectionTableHandle, applyProjectionProjections, applyProjectionAssignments));
```

Comment:
I don't like the fact you inject the expected output and you validate if you get that output. I would prefer if you would define logic for`MockConnectorFactory` and validate if that logic worked. That logic should match the real connector behaviour. 

Otherwise it is like `mockito` which we try to avoid.

So here let's define some rules about what kinds of expresssion are pushable and then write tests for that. For example you can assume that your mock connector can allow to pushdown projects of  `int` literals.

Reply:
@kokosing  Hmm. I think I see your point. The confusion that I have here is that the space of `ProjectionApplicationResult` values that a connector can return varies widely. At this point in the APIs, the connectors do not provide the recipe/rules on creating new handles and assignments, rather, they provide the handles and projections directly.

Also, the API doesn't force connectors to pushdown individual columns, that we can formulate as rules. It all happens at once.  The "logic" in `applyProjection` can be complicated and vary widely for real connectors. For example, say a connector supports dereference pushdown, and returns projected columnhandles in case of a dereference expression like `x.y`. For a column "a" of type `row(b bigint, c bigint)`, if the connector sees that "a.b" and "a.c" *both* are being referenced for a column, it is allowed to choose not to push them down. Because of this possible variety in implementations, I feel that it's better to test the actual implementations, rather than relying on duplicated mock implementations. If we want to test with a rule based approach here, then it might require tightening the `PushProjectionIntoTableScan` API.

That's why I thought that we may want to break this down into two parts. One, test `PushXIntoTableScan` with a "dumb" Mock connector. Secondly, we test the optimizer along with individual connector implementations, by extending something like `ConnectorRuleTester`.  That also helps us "sanity check" things more efficiently as these pushdowns evolve.

Reply:
I have a feeling that this approach of injecting expected values does not help us to test individual connector implementation where you won't be able to inject the expected value. So I would expect that `extending something like ConnectorRuleTester` could be quite difficult.

>  Because of this possible variety in implementations, I feel that it's better to test the actual implementations, rather than relying on duplicated mock implementations. 

I would have just a single implementation for mock. Or just couple of them where which would be abstracted in a way it could be easily reused.